aggregate/tests/test_hail.py

```diff
@@ -0,0 +1,102 @@
+"""Test hail aggregate functionality"""
+import unittest
+from collections import defaultdict
+from unittest.mock import patch
+
+from aggregate.billing_functions.hail import get_finalised_entries_for_batch
+
+
+class TestHailGetFinalisedEntriesForBatch(unittest.TestCase):
+    """Test the batch -> list[entries] functionality"""
+
+    @patch('aggregate.billing_functions.utils.get_currency_conversion_rate_for_time')
+    def test_simple(self, mock_currency_conversion_rate):
```

Comment:
As a suggestion, I like to test one bit of functionality per case where possible. For example I tend to split large tests like this into separate cases with descriptive names, and use the `setUp` instance method create and store mock data available for each test. It makes it easier for me to understand exactly where a function is breaking. However probably not applicable if this is more of an integration test.

Reply:
Haha yes this would be ideal, I sort of incidentally did the hard work when I wrote these tests and evaluated them by hand. There are some doctests littered around (which get picked up by CI), but proper unittests would be good to action in follow-up PRs.